### DISCUSSION TITLE
Can InputSelector choices be loaded lazily?

### DISCUSSION THREAD

**Discussion by haphamdev** [wezterm/wezterm#6310] [Q&A] [answered]
### What Operating System(s) are you running on?

macOS

### WezTerm version

wezterm 20240203-110809-5046fc22

### Ask your question!

For `InputSelector,` can `choices` be assigned to a function that will be invoked lazily when the key mapping is hit?

For example:
```
local get_choices = function()
   -- generate the choices here and return the table
end

--key mapping
{
	key = "t",
	action = act.InputSelector({
		fuzzy = true,
		choices = get_choices,    -- use the function name here
		action = wezterm.action_callback(function (window, pane, id, label) 
			wezterm.log_info(label)
		end)
	}),
},
```

**Comment by haphamdev** [comment]
I can think of many use cases for this:
- Open one of frequently-used directories/projects with fuzzy search
- Switching between tabs with fuzzy search
- Run some frequently-used commands with fuzzy search

**Comment by haphamdev** [comment]
I came up with an idea of using custom event

Map `Leader-e` to publish a custom event
```
{
  key = "e",
  mods = "LEADER",
  action = act.EmitEvent("cev-tab-switching"),
},
```
Event handler: get all tabs as sources for InputSelector
```
wezterm.on("cev-tab-switching", function(win, pane)
  local tabInfo = map(win:mux_window():tabs_with_info(), function(_, tab)
    return { label = tab.tab:get_title(), id = tostring(tab.tab:tab_id()) }
  end)

  win:perform_action(
    act.InputSelector({
      action = wezterm.action_callback(function(w, p, id, label)
        log_info("Tab " .. label .. " was chosen")
      end),
      choices = tabInfo,
      title = "Jump to tab:",
      fuzzy = true,
    }),
    pane
  )
end)

```


The above `map` function is like this
```
local map = function(t, func)
  local result = {}
  for key, value in pairs(t) do
    local v = func(key, value)
    table.insert(result, v)
  end
  return result
end
```

**Comment by haphamdev** [comment]
Btw, I found the action `ShowTabNavigator` but it does not support fuzzy search at first. We have to press `/` to start searching :(

**Comment by wez** [comment, marked as answer]
https://wezfurlong.org/wezterm/config/lua/keyassignment/InputSelector.html#example-of-dynamically-constructing-a-list

**Reply by haphamdev** [reply]
That's not what I mean.
I mean lazy-loading the choices when the key map is hit.

In your example, the choices are populated when the config is loaded.

**Reply by wez** [reply]
Please read the example carefully, it does what you want, you just need to read and understand it.
